engine.go

```diff
@@ -55,6 +55,65 @@ type TemporaryUser struct {
 	Password string
 }
 
+// PreparedDataCache manages all the prepared data for every session for every query for an engine
+type PreparedDataCache struct {
```

Comment:
This is beautiful! It encapsulates this logic cleanly, shrinks the locking granularity, and makes this code a lot easier to read. Nice work! 👏